source/common/http/utility.cc

```diff
@@ -1073,6 +1073,107 @@ std::string Utility::PercentEncoding::decode(absl::string_view encoded) {
   return decoded;
 }
 
+namespace {
+// %-encode all ASCII character codepoints, EXCEPT:
+// ALPHA | DIGIT | * | - | . | _
+// SPACE is encoded as %20, NOT as the + character
+constexpr uint32_t kUrlEncodedCharTable[] = {
+    // control characters
+    0b11111111111111111111111111111111,
+    // !"#$%&'()*+,-./0123456789:;<=>?
+    0b11111111110110010000000000111111,
+    //@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_
+    0b10000000000000000000000000011110,
+    //`abcdefghijklmnopqrstuvwxyz{|}~
+    0b10000000000000000000000000011111,
+    // extended ascii
+    0b11111111111111111111111111111111,
+    0b11111111111111111111111111111111,
+    0b11111111111111111111111111111111,
+    0b11111111111111111111111111111111,
+};
+
+constexpr uint32_t kUrlDecodedCharTable[] = {
+    // control characters
+    0b00000000000000000000000000000000,
+    // !"#$%&'()*+,-./0123456789:;<=>?
+    0b01011111111111111111111111110101,
+    //@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_
+    0b11111111111111111111111111110101,
+    //`abcdefghijklmnopqrstuvwxyz{|}~
+    0b11111111111111111111111111100010,
+    // extended ascii
+    0b00000000000000000000000000000000,
+    0b00000000000000000000000000000000,
+    0b00000000000000000000000000000000,
+    0b00000000000000000000000000000000,
+};
+
+bool testChar(const uint32_t table[8], char c) {
+  uint8_t uc = static_cast<uint8_t>(c);
+  return (table[uc >> 5] & (0x80000000 >> (uc & 0x1f))) != 0;
+}
```

Comment:
Is this an existing algorithm for checking whether a char is one of a well known set of chars? Not immediately obvious to me how it this works. First part takes the most significant few bits and use them to index into one of the bitmasks, and then the least significant bits are compared against the bitmask?

Reply:
Yes, we use it in other places, like UHV. It is CPU cache friendly bit array lookup.

The character validity check could be implemented as a table of size 256 with 0 or 1 in each element indicating character validity. This algorithm packs all 1s and 0s into bits of 32 bit words. That expression with shifts is equivalent to division by 32 and getting the remainder of division by 32 to figure which 32 bit word and which bit in that word to check. It is more CPU cache friendly because table size it 64 bytes instead of 256.

Reply:
Makes sense, thanks for explaining!